compare_platforms/compare_platforms_read_only.c

```diff
@@ -0,0 +1,147 @@
+    
+    review.username = username;
+    review.realname = realname;
+    review.privatereview = privatereview;
+    review.publicreview = publicreview;
+    review.permissions = permissions;
+    print_details(&review);
+
+#ifdef __CHERI_PURE_CAPABILITY__
+    printf("\nsmallsz=%zx, CRRL(smallsz)=%zx\n", smallsz,
+        __builtin_cheri_round_representable_length(smallsz));
+    printf("biggersz=%zx, CRRL(biggersz)=%zx\n", biggersz,
+        __builtin_cheri_round_representable_length(biggersz));
+    printf("username=%#p realname=%#p diff=%tx\n", username, realname, realname - username);
+    printf("realname=%#p privatereview=%#p diff=%tx\n", realname, privatereview, privatereview - realname);
+    printf("privatereview=%#p publicreview=%#p diff=%tx\n", privatereview, publicreview, publicreview - privatereview);
+    printf("publicreview=%#p permissions=%#p diff=%tx\n", publicreview, permissions, permissions - publicreview);
+    printf("realname=%#p permissions=%#p diff=%tx\n", realname, permissions, permissions - realname);
+#else
+    printf("\nusername=%#p realname=%#p diff=%tx\n", username, realname, realname - username);
+    printf("realname=%#p privatereview=%#p diff=%tx\n", realname, privatereview, privatereview - realname);
+    printf("privatereview=%#p publicreview=%#p diff=%tx\n", privatereview, publicreview, publicreview - privatereview);
+    printf("publicreview=%#p permissions=%#p diff=%tx\n", publicreview, permissions, permissions - publicreview);
+    printf("realname=%#p permissions=%#p diff=%tx\n", realname, permissions, permissions - realname);
+#endif
+
+    assert((size_t)(username + smallsz) <= (size_t)realname); 
+    assert((size_t)(privatereview + biggersz) <= (size_t)publicreview);
+    assert((size_t)(realname + smallsz) <= (size_t)permissions); 
+
+// If we are in a CHERI environment, protect the review by setting
+// the relevant struct members (in this case all of them)
+// to read-only.  
+bool b_improved = false;
+  
+#ifdef __CHERI_PURE_CAPABILITY__
+    if(can_write(&review) == false) {
+        printf("\nSetting the review to read-only.\n");
+        struct review *ro_review = set_read_only(&review);
+        assert((cheri_perms_get(ro_review) & CHERI_PERM_STORE) == 0);
+
+        printf("\nThe struct is read-only so trying to change the review will make the program crash...\n");
+        fflush(stdout);
+        
+        b_improved = try_improve_publicreview(ro_review, biggersz, false);
+        
+        // This line should be unreachable.
+        print_details(ro_review);
+    }
+# else
+    // Try editing the public review in a non-CHERI environment,
+    // by simply checking the write permissions and refusing if none are found.
+    b_improved = try_improve_publicreview(&review, biggersz, false);
+#endif
+
+    // This code will be unreachable in a CHERI environment,
+    // but it would crash whether or not the struct had been set to read-only.
```

Comment:
I am assuming it's unreachable due to the read-only aspect. Could we have a switch to turn that off, so that we can see the crash in CHERI as well, when we set the switch?

Reply:
I've started getting an exec format error when I try to run this in morello-purecap. This started after I added a switch, but reverting hasn't fixed it.

Reply:
Added in lines 148-156.

Reply:
Comment header amended to reflect addition of the switch.